p2panda-core/src/serde.rs

```diff
@@ -0,0 +1,603 @@
+
+                let payload_hash: Option<Hash> = match payload_size {
+                    0 => None,
+                    _ => {
+                        let hash: Hash = seq
+                            .next_element()
+                            .map_err(|_| {
+                                SerdeError::custom("invalid payload hash, expected bytes")
+                            })?
+                            .ok_or(SerdeError::custom("payload hash missing"))?;
+                        Some(hash)
+                    }
+                };
+
+                let timestamp: u64 = seq
+                    .next_element()
+                    .map_err(|_| SerdeError::custom("invalid timestamp, expected u64"))?
+                    .ok_or(SerdeError::custom("timestamp missing"))?;
+
+                let seq_num: u64 = seq
+                    .next_element()
+                    .map_err(|_| SerdeError::custom("invalid sequence number, expected u64"))?
+                    .ok_or(SerdeError::custom("sequence number missing"))?;
+
+                let backlink: Option<Hash> = match seq_num {
+                    0 => None,
+                    _ => {
+                        let hash: Hash = seq
+                            .next_element()
+                            .map_err(|err| {
+                                SerdeError::custom(format!(
+                                    "invalid backlink, expected bytes {err}"
+                                ))
+                            })?
+                            .ok_or(SerdeError::custom("backlink missing"))?;
+                        Some(hash)
+                    }
+                };
+
+                let previous: Vec<Hash> = seq
+                    .next_element()
+                    .map_err(|_| SerdeError::custom("invalid previous links, expected array"))?
+                    .ok_or(SerdeError::custom("previous array missing"))?;
+
+                let extension: Option<E> = seq
+                    .next_element()
+                    .map_err(|err| SerdeError::custom(format!("invalid extension: {err}")))?;
+
+                Ok(UnsignedHeader {
+                    version,
+                    public_key,
+                    payload_hash,
+                    payload_size,
+                    timestamp,
+                    seq_num,
+                    backlink,
+                    previous,
+                    extension,
+                })
+            }
+        }
+
+        deserializer.deserialize_seq(HeaderVisitor::<E> {
+            _marker: PhantomData,
+        })
+    }
+}
+
+impl<E> Serialize for Header<E>
+where
+    E: Clone + Serialize + DeserializeOwned,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(2))?;
+
+        seq.serialize_element(&self.0)?;
+        seq.serialize_element(&self.1)?;
```

Comment:
I'm afraid that this changes the operation format now with the signature being parsed first instead of the version.

I think it doesn't matter too much _where_ the signature is parsed, probably it's good early on as it's a mandatory field, but the version should always be first to make sure we support backwards compatibility. The parser should be able to adjust itself right from the beginning, according to the operation version.

A solution could be here to combine the serializer methods of `Header` and `UnsignedHeader` with common logic and something along the lines of `fn serialize_header<S: Blubs>(serializer: S, signature: Option<Signature>)`.

Reply:
Oh damn, yeh of course. I'll have a go at that now.

Reply:
Sounds great! Yeah, `Deserialize` on `UnsignedHeader` is not necessary, makes it easier now ✌️